Support external image viewers

llfw opened the pull request:
- entities: move from_dict to a new module
- Use from_dict to load the settings files.
- settings: add Media configuration
- Add support for external image viewer

Commit:
entities: move from_dict to a new module

This function is generally useful, so move it to tooi.utils.from_dict.

Commit:
Use from_dict to load the settings files.

* Move Configuration from context to settings.
* Move the command-line options into a subclass, Options.
* Use read_dict() to read the entire Configuration class from
  settings.toml at startup.

Commit:
settings: add Media configuration

For now, the only available option is "image_viewer".

Commit:
Add support for external image viewer

The image viewer is configured by the media.image_viewer option.

When viewing media, we download all of the status's attachments to a
temporary directory and invoke the image viewer on the downloaded files,
passing each file as an argv element.  This happens in a @work job, so
the image viewer doesn't block the UI.

Tested with swayimg.

Commit:
TimelineTab: handle case where no image viewer is configured

Show an error modal for this, since the user probably wants to be aware
of the problem and fix it.

Commit:
TimelineTab: don't spawn image viewer if status has no images

ihabunek commented on tooi/settings.py:
```diff
@@ -56,7 +76,7 @@ def _load_settings() -> dict[str, Any]:
 def get_settings():
     settings = _load_settings()
-    return settings.get('options', {})
+    return from_dict(Configuration, settings)
```
This breaks if the settings file does not include both the `[media]` and `[options]` sections.

llfw replied:
fixed in be0cd88

ihabunek committed:
Extract temp file download into a context manager

ihabunek committed:
Keep the extension for downloaded temp files

This might help some image viewers in decoding the image.

ihabunek reviewed (commented):
I found an issue. Additionally, I wanted to extract the downloading logic so i did. :)

Commit:
from_dict: do not call _convert() on default value

If a class member has a default value, assume that the type of the
default value is already appropriate for the member and don't pass it to
_convert().  This avoids an issue if the default value was another
dataclass, which caused an error when _convert tried to call get() on
it.

ghost commented:
rebased on main (e0e034ad565a179ff20188a707a3c367887ccaf7)

Commit:
TimelineTab: fix timeline_refresh option

Caused by incomplete rebase.

ihabunek commented:
Thanks, I've been waiting for this one.
